feat(create-vite): use "solution" tsconfig so that vite.config.ts is type checked

philbates35 opened the pull request:
### Description

Currently when running `pnpm run build` on a fresh Vite project using any of these templates, `tsc` doesn't fail when there are type errors in `vite.config.ts`, such as something obvious like:

```ts
const foo: string = 123;
```

Passing `-b` makes it run for all project references, meaning that `vite.config.ts` starts to be type checked too during build.

In order to not generate a `vite.config.js` and `vite.config.d.ts` we need to make the main `tsconfig.json` have nothing but references to `tsconfig.node.json` and a new `tsconfig.app.json`, as is done in the create-vue project, see:
https://github.com/vuejs/create-vue/blob/f75fd9813a624b8e44b012608335721901aba00b/template/tsconfig/base/tsconfig.json

Note: I am unfamiliar with Svelte which is the only template that I wasn't sure how to apply this to, its script looks like this:

```
"check": "svelte-check --tsconfig ./tsconfig.json"
```
I checked the docs of the `svelte-check` project but I couldn't see anyway to pass options through to `tsc`, if you know of a way to do that then let me know and I'll update the PR.


close #12590 
close #11396

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [X] New Feature
- [ ] Documentation update
- [ ] Other

sapphi-red commented on packages/create-vite/template-preact-ts/package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "dev": "vite",
-    "build": "tsc && vite build",
+    "build": "tsc -b && vite build",
```
This generates `vite.config.js` and `vite.config.d.ts`. So we cannot merge this.

Normally adding `noEmit: true` to `tsconfig.node.json` works but in this case, we cannot simply set that. TypeScript doesn't allow it (`Referenced project 'D:/documents/GitHub/vite/packages/create-vite/template-preact-ts/tsconfig.node.json' may not disable emit.` error happens) (related: https://github.com/microsoft/TypeScript/issues/53979).

We can make it work by adding a tsconfig that references both `tsconfig.app.json` and `tsconfig.node.json` like create-vue does (https://github.com/vuejs/create-vue/pull/274). It seems there's a small problem with that though (https://github.com/vuejs/create-vue/issues/437).

philbates35 replied:
Thanks for pointing that out @sapphi-red, I hadn't realised that it was generating `vite.config.js` and `vite.config.d.ts` :sweat_smile: 

I've updated the PR to do it exactly how its done in https://github.com/vuejs/create-vue/blob/f75fd9813a624b8e44b012608335721901aba00b/template/tsconfig/base/tsconfig.json. Testing locally looks promising - I temporarily added a simple TS error to both a src file and `vite.config.json` and the errors in both show up fine, and only once.

I don't think https://github.com/vuejs/create-vue/issues/437 will be an issue here, from reading that issue it sounds like it happens because the two `tsconfig.*.json` files included the same set of files, whereas here there is no overlap - `tsconfig.app.json` includes only `src` while `tsconfig.node.json` only includes `vite.config.json`.

If you're happy with this, would you be able to let me know what you would like me to do, if anything, with the `svelte-ts` template? I'm unfamiliar with `svelte-check` so I don't know if its even possible to pass `-b` through.

philbates35 replied:
Just pushed an update to set `tsBuildInfoFile` to `node_modules/.tmp/` as its done in create-vue to avoid cluttering up the project root with various `.tsbuildinfo` files, see: https://github.com/vuejs/create-vue/blob/main/template/tsconfig/base/tsconfig.node.json#L13

sapphi-red replied:
> If you're happy with this, would you be able to let me know what you would like me to do, if anything, with the svelte-ts template?

It seems `svelte-check` doesn't support project references. For now, I think running `svelte-check && tsc -p tsconfig.node.json` instead would be fine.

> I don't think vuejs/create-vue#437 will be an issue here.

On second thought, I think this is fine. As long as it's not deduped by TypeScript, we won't be able to solve this. Running two `tsc` (one for app and one for node) alongside would output the error in each, too.

philbates35 replied:
@sapphi-red 

> It seems svelte-check doesn't support project references. For now, I think running svelte-check && tsc -p tsconfig.node.json instead would be fine.

I've pushed the follwing update:

```diff
-    "check": "svelte-check --tsconfig ./tsconfig.json"
+    "check": "svelte-check --tsconfig ./tsconfig.json && tsc -p tsconfig.node.json"
```

> On second thought, I think this is fine. As long as it's not deduped by TypeScript, we won't be able to solve this. Running two tsc (one for app and one for node) alongside would output the error in each, too.

Makes sense :+1:

philbates35 committed:
feat(create-vite): pass -b to tsc so that vite.config.ts is checked

Currently when running "pnpm run build" on a fresh Vite project using
any of these templates, tsc doesn't fail when there are type errors
in vite.config.ts, such as something obvious like:

    const foo: string = 123;

Passing -b makes it run for all project references, meaning that
vite.config.ts starts to be type checked too during build.

In order to not generate a vite.config.js and vite.config.d.ts we
need to make the main tsconfig.json have nothing but references
to tsconfig.node.json and a new tsconfig.app.json, as is done
in the create-vue project, see:
https://github.com/vuejs/create-vue/blob/f75fd9813a624b8e44b012608335721901aba00b/template/tsconfig/base/tsconfig.json

sapphi-red commented:
One thing we might need to consider is that this might make more users to encounter https://github.com/vitejs/vite/issues/15870.

sapphi-red committed:
fix: add missing tsBuildInfoFile option in template-svelte-ts

sapphi-red committed:
fix: add noEmit to tsconfig.node.json so that vite.config.{js,d.ts} is not generated

patak-dev commented:
Let's wait for a tsconfck update before merging this one.

philbates35 commented:
@patak-dev sounds good - just out of interest, which update are we waiting for and what's the reason?

dominikg commented:
why is this turning on composite and storing a file in node_modules/.tmp ? This feels opinionated

cc @dummdidumm regarding svelte-check, iirc it does check all files, not just .svelte.  It should not be needed to chain tsc after it in svelte templates

sapphi-red commented:
> why is this turning on composite

It is to run type check on `vite.config.ts` too. Another way to do this is to call `tsc && tsc -p tsconfig.node.json` instead of `tsc -b`.

> storing a file in node_modules/.tmp

AFAIK when using project references, TypeScript outputs tsbuildinfo files and defaults to the same directory with `tsconfig.json`. There's no way to disable the output.

> Is there a way to make this work that does not require additional files?

We can merge `tsconfig.app.json` and `tsconfig.node.json`, but that allows node-related types to be used in source files and browser-related types to be used in config files (e.g. `process.env` in `src/index.ts`, `document.querySelector` in `vite.config.ts`).

dummdidumm commented:
@dominikg chaining looks fine, assuming the vite config does not import any Svelte code

dominikg commented:
> AFAIK when using project references, TypeScript outputs tsbuildinfo files and defaults to the same directory with `tsconfig.json`. There's no way to disable the output.

you can use references without turning on composite (composite makes incremental true by default https://www.typescriptlang.org/tsconfig#incremental , which generates these files)

calling it tsconfig.node.json is also slightly controversial for users that use other environments such as deno or bun. 

If you have to add a second config file, i'd use a more neutral name and not use composite.  Maybe not even use references but instead add a separate script that just typechecks vite.config.ts with a tsconfig.viteconfig.json.

dominikg commented:
to expand on the separate command part, `build` may not be the right command to typecheck the vite config. The vite configs included in create-vite don't have errors, if and how these are checked as part of build should be decided by the user.

dominikg commented:
enabling composite/incremental can also slow down CI build times and it's not clear if the `node_modules` directory always exists as a sibling of the tsconfig file.  I think this needs to be actively done by the user instead of being default in create-vite.

sapphi-red commented:
> you can use references without turning on composite (composite makes incremental true by default https://www.typescriptlang.org/tsconfig#incremental , which generates these files)

> enabling composite/incremental can also slow down CI build times and it's not clear if the `node_modules` directory always exists as a sibling of the tsconfig file. I think this needs to be actively done by the user instead of being default in create-vite.

I didn't know about that. Then, I think we should remove `composite` and `tsBuildInfoFile` if we'd go this way.

> calling it tsconfig.node.json is also slightly controversial for users that use other environments such as deno or bun.

tsconfig.node.json was there for more than two years. So I want to split that discussion in a different PR.

---

Maybe removing `tsconfig.node.json` and adding `vite.config.ts` to `include` of `tsconfig.json` is better? 🤔 It doesn't have much benefit of having it in the current state (https://github.com/vitejs/vite/issues/11396#issuecomment-1474749697).

bluwy reviewed (approved):
Let's give this a try. I'm also concern that this makes a simple Vite project complicated with having 3 tsconfigs ootb, but we can gauge the perception of this later, and if many others feel the same way, we can simplify as a single tsconfig and document how to typecheck the vite config separately.

fz6m commented:
IMO, making things complex. this shouldn't be included in a **basic** template.

rschristian commented:
Thought it worth mentioning that this does seem to cause errors out-of-the-box in every JSX lib on StackBlitz -- it can't seem to utilize the `"references"` field, so there's errors as TS can't find a `"jsx"` config anywhere.

I imagine this is something StackBlitz has to figure out rather than an issue here, but with StackBlitz being a recommended option for reproductions... something worth considering for the moment, anyways.

patak-dev commented:
There is a fix for StackBlitz that should be deployed soon, so that won't be an issue for people using this scheme. But we are getting other issues as the one that @kasperpeulen has brought up, so I think we should review this change. Would you open an issue so we can better track this @kasperpeulen?

kasperpeulen commented:
@patak-dev here is the issue:
https://github.com/vitejs/vite/issues/17638

fulldecent commented:
This PR broke production builds in a way that the prior version didn't.

---

# Old workflow

Use `yarn dev` on the default `vue-ts` template and do things.

Everything is great. Get ready for bed.

Now run `yarn build` and deploy.

Everything is great. Sleep.

# New workflow

Use `yarn dev` on the default `vue-ts` template and do things.

Everything is great. Get ready for bed.

Now run `yarn build` and all kinds of errors. Go back to step one.

No sleep.
